File: backend/app/api/v1/endpoints/trust_layer.py
# =============================================================================
# Stratum AI - Trust Layer API Router
# =============================================================================
"""
API endpoints for Trust Layer features:
- Signal health monitoring
- Attribution variance tracking
- Trust banners and status
"""

from datetime import date
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.deps import get_current_user
from app.db.session import get_async_session
from app.features.service import can_access_feature
from app.models.trust_layer import FactSignalHealthDaily, SignalHealthStatus
from app.quality.trust_layer_service import (
    AttributionVarianceService,
    SignalHealthService,
)
from app.schemas.response import APIResponse

# SECURITY (STRAT-SC-001/C3): the old per-org request guards were the only
# auth on these routes; deleted in the de-tenanting sweep, so real auth is
# enforced router-wide here.
router = APIRouter(
    tags=["trust-layer"],
    dependencies=[Depends(get_current_user)],
)


# =============================================================================
# Signal Health Endpoints
# =============================================================================


@router.get("/signal-health", response_model=APIResponse[Dict[str, Any]])
async def get_signal_health(
    target_date: Optional[date] = Query(default=None, alias="date"),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get signal health status and metrics for the organization.

    Returns:
    - status: Overall health status (ok/risk/degraded/critical)
    - automation_blocked: Whether automation should be blocked
    - cards: Metric summary cards
    - platform_rows: Per-platform breakdown
    - banners: Trust banners to display
    """
    # Check feature flag
    if not await can_access_feature(db, "signal_health"):
        raise HTTPException(
            status_code=403,
            detail="Signal health feature is not enabled",
        )

    service = SignalHealthService(db)
    data = await service.get_signal_health(target_date)

    return APIResponse(success=True, data=data)


@router.get("/signal-health/history", response_model=APIResponse[Dict[str, Any]])
async def get_signal_health_history(
    days: int = Query(default=7, ge=1, le=30),
    platform: Optional[str] = None,
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get signal health history for the past N days.
    Useful for trend analysis and reporting.
    """
    if not await can_access_feature(db, "signal_health"):
        raise HTTPException(status_code=403, detail="Feature not enabled")

    # For now, return empty history - will be populated by rollup tasks
    return APIResponse(
        success=True,
        data={
            "days": days,
            "platform": platform,
            "history": [],
            "message": "History will be available after daily rollup runs",
        },
    )


# =============================================================================
# Account-Level Signal Health Endpoints
# =============================================================================


@router.get("/signal-health/by-account", response_model=APIResponse[Dict[str, Any]])
async def get_signal_health_by_account(
    target_date: Optional[date] = Query(default=None, alias="date"),
    platform: Optional[str] = Query(default=None, description="Filter by platform"),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get signal health breakdown by ad account.

    Returns per-account signal health metrics grouped by platform and account_id.
    Useful for identifying which specific ad accounts have degraded signals.
    """
    if not await can_access_feature(db, "signal_health"):
        raise HTTPException(
            status_code=403, detail="Signal health feature is not enabled"
        )

    if target_date is None:
        target_date = date.today()

    # Query signal health records grouped by platform + account_id
    conditions = [
        FactSignalHealthDaily.date == target_date,
        FactSignalHealthDaily.account_id.isnot(None),
    ]
    if platform:
        conditions.append(FactSignalHealthDaily.platform == platform)

    result = await db.execute(
        select(FactSignalHealthDaily)
        .where(*conditions)
        .order_by(
            FactSignalHealthDaily.platform,
            FactSignalHealthDaily.account_id,
        )
        .limit(1000)
    )
    records = result.scalars().all()

    # Enrich with account names from AdAccount
    from app.models.campaign_builder import AdAccount

    account_names_result = await db.execute(
        select(
            AdAccount.platform_account_id,
            AdAccount.name,
            AdAccount.business_name,
        )
    )
    account_lookup = {
        row.platform_account_id: {
            "name": row.name,
            "business_name": row.business_name,
        }
        for row in account_names_result.all()
    }

    # Build response
    accounts: List[Dict[str, Any]] = []
    status_counts = {"ok": 0, "risk": 0, "degraded": 0, "critical": 0}

    for record in records:
        account_info = account_lookup.get(record.account_id, {})
        status_val = (
            record.status.value
            if hasattr(record.status, "value")
            else str(record.status)
        )
        status_counts[status_val] = status_counts.get(status_val, 0) + 1

        accounts.append(
            {
                "platform": record.platform,
                "account_id": record.account_id,
                "account_name": account_info.get("name", record.account_id),
                "business_name": account_info.get("business_name"),
                "status": status_val,
                "emq_score": record.emq_score,
                "event_loss_pct": record.event_loss_pct,
                "freshness_minutes": record.freshness_minutes,
                "api_error_rate": record.api_error_rate,
                "issues": record.issues,
                "actions": record.actions,
                "notes": record.notes,
            }
        )

    # Determine overall status
    if status_counts["critical"] > 0:
        overall_status = "critical"
    elif status_counts["degraded"] > 0:
        overall_status = "degraded"
    elif status_counts["risk"] > 0:
        overall_status = "risk"
    else:
        overall_status = "ok"

    return APIResponse(
        success=True,
        data={
            "date": target_date.isoformat(),
            "overall_status": overall_status,
            "status_counts": status_counts,
            "total_accounts": len(accounts),
            "accounts": accounts,
        },
    )


# =============================================================================
# Attribution Variance Endpoints
# =============================================================================


@router.get("/attribution-variance", response_model=APIResponse[Dict[str, Any]])
async def get_attribution_variance(
    target_date: Optional[date] = Query(default=None, alias="date"),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get attribution variance between platform and GA4.

    Returns:
    - status: Variance status (healthy/minor/moderate/high)
    - overall_revenue_variance_pct: Overall revenue variance percentage
    - cards: Summary metric cards
    - platform_rows: Per-platform breakdown
    - banners: Attribution variance banners
    """
    if not await can_access_feature(db, "attribution_variance"):
        raise HTTPException(
            status_code=403,
            detail="Attribution variance feature is not enabled",
        )

    service = AttributionVarianceService(db)
    data = await service.get_attribution_variance(target_date)

    return APIResponse(success=True, data=data)


# =============================================================================
# Combined Trust Status Endpoint
# =============================================================================


@router.get("/trust-status", response_model=APIResponse[Dict[str, Any]])
async def get_trust_status(
    target_date: Optional[date] = Query(default=None, alias="date"),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get combined trust status including signal health and attribution variance.
    This is the main endpoint for the Trust Layer banner/status.

    Returns:
    - overall_status: Combined trust status
    - automation_allowed: Whether automation can proceed
    - signal_health: Signal health summary (if enabled)
    - attribution_variance: Attribution variance summary (if enabled)
    - banners: Combined banners to display
    """
    result = {
        "date": (target_date or date.today()).isoformat(),
        "overall_status": "ok",
        "automation_allowed": True,
        "signal_health": None,
        "attribution_variance": None,
        "banners": [],
    }

    # Get signal health if enabled
    if await can_access_feature(db, "signal_health"):
        service = SignalHealthService(db)
        signal_data = await service.get_signal_health(target_date)
        result["signal_health"] = signal_data

        if signal_data["status"] in ["degraded", "critical"]:
            result["overall_status"] = signal_data["status"]
            result["automation_allowed"] = False
            result["banners"].extend(signal_data["banners"])
        elif signal_data["status"] == "risk" and result["overall_status"] == "ok":
            result["overall_status"] = "risk"
            result["banners"].extend(signal_data["banners"])

    # Get attribution variance if enabled
    if await can_access_feature(db, "attribution_variance"):
        service = AttributionVarianceService(db)
        attr_data = await service.get_attribution_variance(target_date)
        result["attribution_variance"] = attr_data

        # Add attribution banners (don't block automation for attribution issues)
        result["banners"].extend(attr_data["banners"])

    return APIResponse(success=True, data=result)


# =============================================================================
# Trust Gate — the real stratum/core decision engine
# =============================================================================
# The TrustGate/SignalHealthCalculator pair in app/stratum/core is the
# documented Trust Engine (CLAUDE.md "Trust Engine Rules") but was never
# reachable from the API. These endpoints adapt the organization's rolled-up
# FactSignalHealthDaily rows into the engine's SignalHealth model and expose
# gate decisions, thresholds, and the derived autopilot mode.


class TrustGateEvaluateRequest(BaseModel):
    """Request to evaluate one automation action against the trust gate."""

    action_type: str = Field(..., description="e.g. increase_budget, pause_campaign")
    entity_type: str = Field("campaign", description="campaign, adset, or ad")
    entity_id: str = Field(..., description="Platform entity ID")
    platform: str = Field("meta", description="meta, google, tiktok, snapchat")
    account_id: str = Field("", description="Platform ad account ID")
    parameters: Dict[str, Any] = Field(default_factory=dict)


async def _latest_health_rows(
    db: AsyncSession, target_date: Optional[date]
) -> List[FactSignalHealthDaily]:
    """Fetch the organization's platform health rows for the most recent
    rollup date at or before ``target_date`` (default today)."""
    upper = target_date or date.today()
    latest = await db.execute(
        select(func.max(FactSignalHealthDaily.date)).where(
            FactSignalHealthDaily.date <= upper,
        )
    )
    as_of = latest.scalar_one_or_none()
    if as_of is None:
        return []

    result = await db.execute(
        select(FactSignalHealthDaily).where(
            FactSignalHealthDaily.date == as_of,
        )
    )
    return list(result.scalars().all())


def _signal_health_from_rows(rows: List[FactSignalHealthDaily]):
    """Adapt rolled-up daily rows into the engine's SignalHealth model.

    Uses the stratum SignalHealthCalculator so component weighting and
    status thresholds live in ONE place. With no rows, returns a zeroed
    critical SignalHealth — the safety default is to block automation on
    unknown data quality, never to wave it through.
    """
    from datetime import datetime, timedelta, timezone

    from app.stratum.core.signal_health import SignalHealthCalculator
    from app.stratum.models import SignalHealth

    if not rows:
        return SignalHealth(
            overall_score=0.0,
            emq_score=0.0,
            freshness_score=0.0,
            variance_score=0.0,
            anomaly_score=0.0,
            status="critical",
            issues=["No signal health data available — run the rollup"],
        )

    emq_scores = [
        row.emq_score / 10.0 for row in rows if row.emq_score is not None
    ]  # calculator expects the platform 0-10 EMQ scale

    freshness_values = [
        row.freshness_minutes for row in rows if row.freshness_minutes is not None
    ]
    last_data_received = (
        datetime.now(timezone.utc) - timedelta(minutes=max(freshness_values))
        if freshness_values
        else None
    )

    calculator = SignalHealthCalculator()
    return calculator.calculate(
        emq_scores=emq_scores or None,
        last_data_received=last_data_received,
    )


async def _trust_gate_for_org(db: AsyncSession):
    """Build a TrustGate using the organization's configured thresholds
    (TRUST-007).

    Reads the org-level thresholds onboarding stored on
    ``organization.settings`` (falling back to the global defaults) instead
    of always using the defaults.
    """
    from app.base_models import get_organization
    from app.stratum.core.trust_gate import TrustGate, TrustGateConfig

    org = await get_organization(db)
    return TrustGate(TrustGateConfig.from_org_settings(org.settings))


@router.get("/trust-gate", response_model=APIResponse[Dict[str, Any]])
async def get_trust_gate_status(
    target_date: Optional[date] = Query(default=None, alias="date"),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get the trust gate's current posture.

    Returns the composite signal health (engine-weighted), the configured
    gate thresholds, the derived autopilot mode, and which automation
    action types are currently allowed vs restricted.
    """
    from app.stratum.core.trust_gate import TrustGate, get_autopilot_mode

    rows = await _latest_health_rows(db, target_date)
    health = _signal_health_from_rows(rows)

    gate = await _trust_gate_for_org(db)
    allowed, restricted = gate.get_allowed_actions(health)
    mode, mode_reason = get_autopilot_mode(health)

    return APIResponse(
        success=True,
        data={
            "data_available": bool(rows),
            "as_of_date": rows[0].date.isoformat() if rows else None,
            "signal_health": {
                "score": health.overall_score,
                "status": health.status,
                "components": {
                    "emq": health.emq_score,
                    "freshness": health.freshness_score,
                    "variance": health.variance_score,
                    "anomaly": health.anomaly_score,
                },
                "issues": health.issues,
            },
            "autopilot_mode": mode,
            "mode_reason": mode_reason,
            "allowed_actions": sorted(allowed),
            "restricted_actions": sorted(restricted),
            "thresholds": gate.get_stats(),
        },
    )


@router.post("/trust-gate/evaluate", response_model=APIResponse[Dict[str, Any]])
async def evaluate_trust_gate(
    body: TrustGateEvaluateRequest,
    db: AsyncSession = Depends(get_async_session),
):
    """
    Evaluate a proposed automation action against the trust gate.

    Returns PASS / HOLD / BLOCK with the engine's reasoning — the same
    decision contract the autopilot uses, exposed for the dashboard's
    "would this run right now?" preview.
    """
    from app.stratum.core.trust_gate import TrustGate
    from app.stratum.models import AutomationAction, Platform

    try:
        platform = Platform(body.platform)
    except ValueError:
        raise HTTPException(
            status_code=400, detail=f"Unknown platform: {body.platform}"
        )

    rows = await _latest_health_rows(db, None)
    health = _signal_health_from_rows(rows)

    action = AutomationAction(
        platform=platform,
        account_id=body.account_id,
        entity_type=body.entity_type,
        entity_id=body.entity_id,
        action_type=body.action_type,
        parameters=body.parameters,
    )

    gate = await _trust_gate_for_org(db)
    result = gate.evaluate(health, action)
    payload = result.to_dict()
    payload["data_available"] = bool(rows)

    return APIResponse(success=True, data=payload)

File: frontend/src/api/trustLayer.ts
/**
 * ADs Growth System - Trust Layer API Hooks
 *
 * React Query hooks for Trust Layer features:
 * - Signal health monitoring
 * - Attribution variance tracking
 * - Combined trust status
 */

import { useMutation, useQuery } from '@tanstack/react-query'
import { apiClient } from './client'

// =============================================================================
// Types
// =============================================================================

export type SignalHealthStatus = 'ok' | 'risk' | 'degraded' | 'critical' | 'no_data'
export type AttributionVarianceStatus = 'healthy' | 'minor_variance' | 'moderate_variance' | 'high_variance' | 'no_data'

export interface MetricCard {
  title: string
  value: string
  status: 'ok' | 'risk' | 'degraded' | 'neutral'
  description?: string
}

export interface TrustBanner {
  type: 'info' | 'warning' | 'error'
  title: string
  message: string
  actions: string[]
}

export interface PlatformHealthRow {
  platform: string
  account_id?: string
  emq_score?: number
  event_loss_pct?: number
  freshness_minutes?: number
  api_error_rate?: number
  status: SignalHealthStatus
}

export interface SignalHealthData {
  date: string
  status: SignalHealthStatus
  automation_blocked: boolean
  cards: MetricCard[]
  platform_rows: PlatformHealthRow[]
  banners: TrustBanner[]
  issues: string[]
  actions: string[]
}

export interface PlatformVarianceRow {
  platform: string
  ga4_revenue: number
  platform_revenue: number
  revenue_delta_pct: number
  ga4_conversions: number
  platform_conversions: number
  conversion_delta_pct: number
  confidence: number
  status: AttributionVarianceStatus
}

export interface AttributionVarianceData {
  date: string
  status: AttributionVarianceStatus
  overall_revenue_variance_pct: number
  overall_conversion_variance_pct: number
  cards: MetricCard[]
  platform_rows: PlatformVarianceRow[]
  banners: TrustBanner[]
}

export interface TrustStatusData {
  date: string
  overall_status: SignalHealthStatus
  automation_allowed: boolean
  signal_health: SignalHealthData | null
  attribution_variance: AttributionVarianceData | null
  banners: TrustBanner[]
}

// --- Trust gate (the real engine, exposed via /trust-gate) ---

export type GateDecision = 'pass' | 'hold' | 'block'
export type AutopilotGateMode = 'normal' | 'limited' | 'cuts_only' | 'frozen'

export interface GateSignalHealth {
  score: number
  status: string
  components: Record<string, number>
  issues: string[]
}

export interface TrustGateThresholds {
  pass_threshold: number
  hold_threshold: number
  high_risk_threshold: number
  conservative_threshold: number
  high_risk_actions: string[]
  conservative_actions: string[]
  always_allowed_actions: string[]
}

export interface TrustGateStatusData {
  data_available: boolean
  as_of_date: string | null
  signal_health: GateSignalHealth
  autopilot_mode: AutopilotGateMode
  mode_reason: string
  allowed_actions: string[]
  restricted_actions: string[]
  thresholds: TrustGateThresholds
}

export interface TrustGateEvaluateInput {
  action_type: string
  entity_type?: string
  entity_id: string
  platform?: string
  account_id?: string
  parameters?: Record<string, unknown>
}

export interface TrustGateEvaluation {
  decision: GateDecision
  signalHealth: GateSignalHealth & { hasCdpData: boolean | null }
  action: { type: string; entity: string; platform: string }
  reason: string
  allowedActions: string[]
  restrictedActions: string[]
  recommendations: string[]
  evaluatedAt: string
  data_available: boolean
}

// =============================================================================
// Hooks
// =============================================================================

/**
 * Fetch signal health data for a tenant.
 */
export function useSignalHealth(tenantId: number, date?: string) {
  return useQuery({
    queryKey: ['signal-health', tenantId, date],
    queryFn: async () => {
      const params = date ? `?date=${date}` : ''
      const response = await apiClient.get<{ data: SignalHealthData }>(
        `/trust/signal-health${params}`
      )
      return response.data.data
    },
    enabled: !!tenantId,
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 5 * 60 * 1000, // Refetch every 5 minutes
  })
}

/**
 * Fetch signal health history for trend analysis.
 */
export function useSignalHealthHistory(
  tenantId: number,
  days: number = 7,
  platform?: string
) {
  return useQuery({
    queryKey: ['signal-health-history', tenantId, days, platform],
    queryFn: async () => {
      const params = new URLSearchParams({ days: days.toString() })
      if (platform) params.append('platform', platform)

      const response = await apiClient.get<{ data: { history: SignalHealthData[] } }>(
        `/trust/signal-health/history?${params}`
      )
      return response.data.data.history
    },
    enabled: !!tenantId,
    staleTime: 5 * 60 * 1000, // 5 minutes
  })
}

/**
 * Fetch attribution variance data for a tenant.
 */
export function useAttributionVariance(tenantId: number, date?: string) {
  return useQuery({
    queryKey: ['attribution-variance', tenantId, date],
    queryFn: async () => {
      const params = date ? `?date=${date}` : ''
      const response = await apiClient.get<{ data: AttributionVarianceData }>(
        `/trust/attribution-variance${params}`
      )
      return response.data.data
    },
    enabled: !!tenantId,
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 5 * 60 * 1000, // Refetch every 5 minutes
  })
}

/**
 * Fetch combined trust status for a tenant.
 * This is the main endpoint for the Trust Layer banner.
 */
export function useTrustStatus(tenantId: number, date?: string) {
  return useQuery({
    queryKey: ['trust-status', tenantId, date],
    queryFn: async () => {
      const params = date ? `?date=${date}` : ''
      const response = await apiClient.get<{ data: TrustStatusData }>(
        `/trust/trust-status${params}`
      )
      return response.data.data
    },
    enabled: !!tenantId,
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 5 * 60 * 1000, // Refetch every 5 minutes
  })
}

/**
 * Fetch the trust gate's current posture: engine-weighted signal health,
 * derived autopilot mode, and which action types are allowed vs restricted.
 */
export function useTrustGateStatus(tenantId: number, date?: string) {
  return useQuery({
    queryKey: ['trust-gate', tenantId, date],
    queryFn: async () => {
      const params = date ? `?date=${date}` : ''
      const response = await apiClient.get<{ data: TrustGateStatusData }>(
        `/trust/trust-gate${params}`
      )
      return response.data.data
    },
    enabled: !!tenantId,
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 5 * 60 * 1000, // Refetch every 5 minutes
  })
}

/**
 * Evaluate a proposed automation action against the trust gate —
 * the dashboard's "would this run right now?" preview. Read-only on the
 * backend: it renders a PASS / HOLD / BLOCK decision, nothing executes.
 */
export function useEvaluateTrustGate(_tenantId: number) {
  return useMutation({
    mutationFn: async (input: TrustGateEvaluateInput) => {
      const response = await apiClient.post<{ data: TrustGateEvaluation }>(
        `/trust/trust-gate/evaluate`,
        input
      )
      return response.data.data
    },
  })
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Get status color for UI rendering.
 */
export function getStatusColor(status: SignalHealthStatus | AttributionVarianceStatus): string {
  const colors: Record<string, string> = {
    ok: 'green',
    healthy: 'green',
    risk: 'yellow',
    minor_variance: 'yellow',
    degraded: 'orange',
    moderate_variance: 'orange',
    critical: 'red',
    high_variance: 'red',
    no_data: 'gray',
  }
  return colors[status] || 'gray'
}

/**
 * Get status label for display.
 */
export function getStatusLabel(status: SignalHealthStatus | AttributionVarianceStatus): string {
  const labels: Record<string, string> = {
    ok: 'Healthy',
    healthy: 'Healthy',
    risk: 'At Risk',
    minor_variance: 'Minor Variance',
    degraded: 'Degraded',
    moderate_variance: 'Moderate Variance',
    critical: 'Critical',
    high_variance: 'High Variance',
    no_data: 'No Data',
  }
  return labels[status] || status
}

/**
 * Get banner icon based on type.
 */
export function getBannerIcon(type: 'info' | 'warning' | 'error'): string {
  const icons: Record<string, string> = {
    info: 'ℹ️',
    warning: '⚠️',
    error: '🚨',
  }
  return icons[type] || 'ℹ️'
}
